### PR TITLE
Support Iceberg version travel by reference name

### PR DESCRIPTION
**Pull request by jackye1995**
## Description

Support Iceberg version travel by reference name.

## Additional context and related issues

Example:

```
SELECT * FROM table AS OF VERSION 'tag-name'
```

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
(x) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

### CODE REVIEW

**Review by alexjo2144 — APPROVED** [review]
Nice

**Comment by alexjo2144** [comment]
Does Spark SQL have syntax for adding a branch/tag? Would also be nice to have a test in `TestIcebergSparkCompatibility`

**Comment by findinpath** [comment]
> Does Spark SQL have syntax for adding a branch/tag? Would also be nice to have a test in TestIcebergSparkCompatibility

See WIP https://github.com/apache/iceberg/pull/6575

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergV2.java [hunk 1]
```diff
@@ -473,6 +473,29 @@ public void testDeletingEntirePartitionedTable()
...
+                .createTag("test-tag", refSnapshotId)
+                .createBranch("test-branch", refSnapshotId)
+                .commit();
+
+        assertUpdate("INSERT INTO " + tableName + " SELECT * FROM tpch.tiny.nation LIMIT 5", 5);
+        assertQuery("SELECT * FROM " + tableName + " FOR VERSION AS OF " + refSnapshotId,
+                "SELECT * FROM nation");
+        assertQuery("SELECT * FROM " + tableName + " FOR VERSION AS OF 'test-tag'",
+                "SELECT * FROM nation");
+        assertQuery("SELECT * FROM " + tableName + " FOR VERSION AS OF 'test-branch'",
```
Comment (findinpath) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergV2.java:493]: I was looking over the following test

https://github.com/apache/iceberg/blob/e56c1cf8c16f527bf91b74399602138e215f4186/core/src/test/java/org/apache/iceberg/TestFastAppend.java#L525-L532

The newly introduced syntax would work to reference the latest snapshot in a branch.
How would we want to reference a specific snapshot from within a branch?

Comment (jackye1995) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergV2.java:493]: > How would we want to reference a specific snapshot from within a branch?

If you want to reference a specific snapshot by ID, then you don't really need to specify the branch, the snapshot ID already uniquely identifies it.

If you want to do time travel within a branch, then yes it is not covered by this PR. We can use some way like `FOR VERSION AS OF 'branch@timestamp'`, but that feels a bit too hacky and requires further discussion. I am working on introducing a syntax like `SELECT * FROM table_name BRANCH branch_name` so you can do `SELECT * FROM table_name BRANCH branch_name FOR TIMESTAMP AS OF xxxx`. We can see how people like that once the PR is published.

**Comment by jackye1995** [comment]
@alexjo2144 @findinpath @findepi I have merged the corresponding PR in Iceberg for Spark: https://github.com/apache/iceberg/pull/6575

Please let me know if this one is good to go. We probably cannot do a Spark compatibility test until the next version of Iceberg release.

**Comment by alexjo2144** [comment]
> We probably cannot do a Spark compatibility test until the next version of Iceberg release.

Makes sense. Mind just creating a GH issue on the project to remind us to do that after upgrading the Iceberg Spark version?

@findepi or @ebyhr need to do a final review but this is good to go as far as I'm concerned.

**Review by findinpath — APPROVED** [review]

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergV2.java [hunk 2]
```diff
@@ -473,6 +473,29 @@ public void testDeletingEntirePartitionedTable()
         assertThat(this.loadTable(tableName).newScan().planFiles()).hasSize(0);
     }
 
+    @Test
+    public void testReadingSnapshotReference()
+    {
+        String tableName = "test_reading_snapshot_reference" + randomNameSuffix();
+        assertUpdate("CREATE TABLE " + tableName + " WITH (partitioning = ARRAY['regionkey']) AS SELECT * FROM tpch.tiny.nation", 25);
+        Table icebergTable = this.loadTable(tableName);
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergV2.java:481]: nit: Remove redundant `this.`. I know some existing tests contain it though.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergV2.java [hunk 3]
```diff
@@ -473,6 +473,29 @@ public void testDeletingEntirePartitionedTable()
...
+                .commit();
+
+        assertUpdate("INSERT INTO " + tableName + " SELECT * FROM tpch.tiny.nation LIMIT 5", 5);
+        assertQuery("SELECT * FROM " + tableName + " FOR VERSION AS OF " + refSnapshotId,
+                "SELECT * FROM nation");
+        assertQuery("SELECT * FROM " + tableName + " FOR VERSION AS OF 'test-tag'",
+                "SELECT * FROM nation");
+        assertQuery("SELECT * FROM " + tableName + " FOR VERSION AS OF 'test-branch'",
+                "SELECT * FROM nation");
+        assertQueryFails("SELECT * FROM " + tableName + " FOR VERSION AS OF 'test-wrong-ref'",
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergV2.java:495]: Can we add another `assertQueryFails` with uppercase of existing reference names? e.g. `TEST-TAG`

**Review by ebyhr — APPROVED** [review]
Did you skip the documentation intentionally because we haven't yet provided the relevant system table (#15649)?

**Comment by marton-bod** [comment]
@jackye1995 Thank you for working on this, this is great!
Just wanted to check if we're also planning to add support for creating branches/tags from Trino? Not sure if any of the engines have added support for branch/tag creation yet

**Comment by jackye1995** [comment]
> if we're also planning to add support for creating branches/tags from Trino? Not sure if any of the engines have added support for branch/tag creation yet

We are working on merging related PRs in Iceberg Spark to add Spark SQL

https://github.com/apache/iceberg/pull/6617
https://github.com/apache/iceberg/pull/6638
https://github.com/apache/iceberg/pull/6637

But the Java API is already there and can be used to do that.

I am not sure how it could be done in Trino, maybe we could add some procedures.

**Comment by bitsondatadev** [comment]
@ebyhr is this ready to be merged?

**Comment by ebyhr** [comment]
@bitsondatadev There're remaining comments and it would be nice to merge #15649 first. Otherwise, users can't know tag and branch names for time travel.

**Comment by jackye1995** [comment]
@ebyhr @bitsondatadev I've updated the PR, let me know what else is needed, thanks.

**Comment by ebyhr** [comment]
Could you rebase on master to resolve conflicts?

**Comment by jackye1995** [comment]
Can I get a confirmation if this PR is ever gonna be merged once rebased? If not I will just go ahead to apply the patch internally and you can close this PR instead of me continuing to rebase.

**Comment by jackye1995** [comment]
Close PR, this will be available in AWS Athena by 3/31 in all AWS regions.

**Comment by ebyhr** [comment]
Superseded by #19111

**Pull request closed without merging**